### PR TITLE
New module vmware_deploy_ovf

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY
Adds a module for deploying a VMWare VM from an OVF file

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
vmware_deploy_ovf

##### ANSIBLE VERSION

```
devel
```

##### ADDITIONAL INFORMATION

Current Caveats:

1. ~~No check mode~~
1. ~~No idempotency~~
1. ~~No OVA support~~
1. ~~Has not been tested on vsphere/vcenter only stand alone esxi~~

### CODE REVIEW

**Comment by dav1x** [comment]
I'm having some issues getting this to work.

```
---
- hosts: localhost
  gather_facts: No
  connection: local

  tasks:
  - name: create the vm
    vmware_deploy_ovf:
      hostname: 10.x.x.25
      datacenter: Boston
      datastore: ose3-vmware-prod
      #username: root
      username: administrator@vsphere.local
      password: ****
      wait_for_ip_address: True
      ovf: ./rhel-atomic-cloud-7.2-10.x86_64.vsphere.ovf                                                                                                                                                                          
    register: deploy

The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_UtK8gU/ansible_module_vmware_deploy_ovf.py", line 354, in <module>
    main()
  File "/tmp/ansible_UtK8gU/ansible_module_vmware_deploy_ovf.py", line 296, in main
    spec_params
  File "build/bdist.linux-x86_64/egg/pyVmomi/VmomiSupport.py", line 580, in <lambda>
  File "build/bdist.linux-x86_64/egg/pyVmomi/VmomiSupport.py", line 386, in _InvokeMethod
  File "build/bdist.linux-x86_64/egg/pyVmomi/SoapAdapter.py", line 1319, in InvokeMethod
  File "build/bdist.linux-x86_64/egg/pyVmomi/SoapAdapter.py", line 938, in SerializeRequest
  File "build/bdist.linux-x86_64/egg/pyVmomi/SoapAdapter.py", line 180, in _SerializeToUnicode
  File "build/bdist.linux-x86_64/egg/pyVmomi/SoapAdapter.py", line 255, in Serialize
  File "build/bdist.linux-x86_64/egg/pyVmomi/SoapAdapter.py", line 449, in _Serialize
  File "/usr/lib64/python2.7/encodings/utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode byte 0x80 in position 5140: invalid start byte

fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_UtK8gU/ansible_module_vmware_deploy_ovf.py\", line 354, in <module>\n    main()\n  File \"/tmp/ansible_UtK8gU/ansible_module_vmware_deploy_ovf.py\", line 296, in main\n    spec_params\n  File \"build/bdist.linux-x86_64/egg/pyVmomi/VmomiSupport.py\", line 580, in <lambda>\n  File \"build/bdist.linux-x86_64/egg/pyVmomi/VmomiSupport.py\", line 386, in _InvokeMethod\n  File \"build/bdist.linux-x86_64/egg/pyVmomi/SoapAdapter.py\", line 1319, in InvokeMethod\n  File \"build/bdist.linux-x86_64/egg/pyVmomi/SoapAdapter.py\", line 938, in SerializeRequest\n  File \"build/bdist.linux-x86_64/egg/pyVmomi/SoapAdapter.py\", line 180, in _SerializeToUnicode\n  File \"build/bdist.linux-x86_64/egg/pyVmomi/SoapAdapter.py\", line 255, in Serialize\n  File \"build/bdist.linux-x86_64/egg/pyVmomi/SoapAdapter.py\", line 449, in _Serialize\n  File \"/usr/lib64/python2.7/encodings/utf_8.py\", line 16, in decode\n    return codecs.utf_8_decode(input, errors, True)\nUnicodeDecodeError: 'utf8' codec can't decode byte 0x80 in position 5140: invalid start byte\n", 
    "module_stdout": "", 
    "rc": 0
}

MSG:

MODULE FAILURE

```

**Comment by sivel** [comment]
@dav1x Is it possible to see the contents of `rhel-atomic-cloud-7.2-10.x86_64.vsphere.ovf`?

That error is coming from a SOAP call in pyvmomi, and since it works for me with my test, and I don't see any strange characters in your task, I'm assuming it is from the `ovf`.

Also, just in case, is it really an OVF?  Or is it an OVA (binary which includes the OVF and the VMDK)?   Currently, this module only supports separate OVF and VMDK, it does not support OVA.

**Comment by dav1x** [comment]
@sivel Sure. 

```
(ose3-ansible-host)(root)(/root/vmware_guest/features/image)(09/13/17) (04:40:20 PM)
-> tar xvvf rhel-atomic-cloud-7.2-10.x86_64.vsphere.ovf 
-rw-r--r-- root/root      3703 2015-11-10 13:20 desc.ovf
-rwxr-xr-x root/root 429120000 2015-11-10 13:20 disk.img
(ose3-ansible-host)(root)(/root/vmware_guest/features/image)(09/13/17) (04:40:27 PM)
-> ls
desc.ovf  disk.img  rhel-atomic-cloud-7.2-10.x86_64.vsphere.ovf
(ose3-ansible-host)(root)(/root/vmware_guest/features/image)(09/13/17) (04:40:28 PM)
-> file desc.ovf 
desc.ovf: ASCII text, with very long lines, with no line terminators
(ose3-ansible-host)(root)(/root/vmware_guest/features/image)(09/13/17) (04:40:39 PM)
-> file disk.img 
disk.img: VMware4 disk image
```
Should I deploy a different image?

**Comment by sivel** [comment]
You should be able to use that, but you will need the separate OVF and VMDK files.

The OVF must only be the XML for the OVF, and the VMDK should match the OVF; for example:

```
  <References>
    <File ovf:href="ubuntu-16.04-amd64-disk1.vmdk" ovf:id="file1" ovf:size="462418944"/>
  </References>
```

In my case the OVF is `ubuntu-16.04-amd64.ovf` and then it references `ubuntu-16.04-amd64-disk1.vmdk`

The VMDK is inferred from the OVF, and read from the same directory as the OVF.

So you would just need to point to your `desc.ovf`.  And assuming that your OVF references `disk.img` and `disk.img` is a VMDK, I assume it should work then.

OVA file support can potentially come later, but the API doesn't support it, so the code would have to deconstruct it, before deploying.

**Comment by sivel** [comment]
I'm going to spend some time on OVA support as well.  I have some working code, I just need to clean things up.  As the code is growing things are getting a little messy.

**Comment by dav1x** [comment]
Ack @sivel 
This worked. I didn't specify a name and the 
```
<VirtualSystem ovf:id="407f1ccd-7a36-4dd6-9f7e-feebec6c73c2">
```
was chosen as the VM name. You said that there was no idempotency and my second run of the playbook agrees with that. The playbook run never errors out after vCenter gives the error that the VM already exists.

**Comment by sivel** [comment]
@dav1x are you sure it actually errored in vCenter? As I understand it, the VM name is not required to be unique, and in my tests running repeatedly just creates multiple VMs with the same name.  They are all given unique IDs.

My OVF defines:

```
<VirtualSystem ovf:id="ubuntu-16.04-amd64">
```

And I just end up with multiple of that name:

```
$ ./govc_darwin_amd64 find -name ubuntu-16.04-amd64
/ha-datacenter/vm/ubuntu-16.04-amd64
/ha-datacenter/vm/ubuntu-16.04-amd64
/ha-datacenter/vm/ubuntu-16.04-amd64
```

File: lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py]: Could you please move these function to vmware module utils so that other modules can also use them ?

- find_resource_pool_by_name
- find_network_by_name
- wait_for_vm_ip

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py
Comment (sivel) [reply lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py]: Yeah, that is my plan.  Currently, as this is a WIP, I am trying to keep things consolidated.

Also, due to the `find_*_by_name` functions here and in module utils being so similar, I may create a base function that does the similar work.

**Comment by sivel** [comment]
OVA support has been added.

**Comment by sivel** [comment]
@jctanner I've copied `wait_for_vm_ip` over from `vmware_guest.py` into `ansible.module_utils.vmware` (but not replaced it's use in `vmware_guest.py`).

Do I have your blessing to license that code as BSD?

https://github.com/ansible/ansible/commit/2c17f9becda010188e57f1751bfcbdf64a600ff8

Update: Wait, never mind.  `ansible.module_utils.vmware` is GPL licensed it seems...

File: lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py
Comment (sivel) [reply lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py]: Done!

**Comment by sivel** [comment]
I'm not positive on where to go from here.  I need to figure out what do to with idempotency.  In my use case, I don't care what the name is, and if I call the module I want a new host.

The `vmware_guest` module uses `name` and `name_match` (first/last) for this, I don't know whether to stick with that, which will causes issues with my use case, or to go a different route.

Coming from having written other cloud modules (rax*) we used things like `exact_count` and a group tag/metadata, but I don't think that is the exact approach I wanted to take now, although it could go there later.

I think I am instead leaning towards some form of "tag" through custom values to determine idempotency.

Any feedback?

**Review by Akasurde — APPROVED** [review]
LGTM

**Comment by sivel** [comment]
As an update, I'm waiting on vcsim support so I can add some tests.  This has been requested in https://github.com/vmware/govmomi/issues/963

**Comment by DanyC97** [comment]
@sivel great work adding this long standing feature.

I'm a bit confused by your last comment, if i got it right:

> waiting on vcsim support 

is this only for testing framework due to  `govmomi` being used instead of `pyvmomi`?

On a different note, could you please confirm if the below e2e (which might include this module + the other `vmware_*` modules ) use case was considered?

* upload the ova file 
* once uploaded convert it to a template
* create VMs from template
     * including disk re-size/ adding new disk etc 

cheers

**Comment by DanyC97** [comment]
@sivel fyi in case there is a need for inspiration [here](https://github.com/vmware/pyvmomi-community-samples/blob/master/samples/deploy_ova.py )

**Comment by sivel** [comment]
@DanyC97

> I'm a bit confused by your last comment, if i got it right:
> 
> waiting on vcsim support
> is this only for testing framework due to govmomi being used instead of pyvmomi?

This is only for testing. We use [vcsim](https://github.com/vmware/govmomi/tree/master/vcsim) as an integration point for our testing. We've decided to hold off, at least for now on merging this in, due to the variety of testing scenarios.  vcsim would help us ensure that things work as expected.

> On a different note, could you please confirm if the below e2e (which might include this module + the other vmware_* modules ) use case was considered?

I can't say specifically that it was, but this module would only target "upload the ova file".  You would need to look to other modules for the additional functionality.

> fyi in case there is a need for inspiration here (https://github.com/vmware/pyvmomi-community-samples/blob/master/samples/deploy_ova.py)

This module supports all of the functionality provided by that script.  In addition to ovf+vmdk (not part of an ova)

**Comment by DanyC97** [comment]
thanks for the info @sivel ! 
this will be a massive step forward once it lands, i hope soon.

File: lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py
Comment (sivel) [inline lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py]: There might be a bug calculating progress here.  I may need to create all the uploaders in the for loop but not start them, calculate total size, and then process the uploaders in their own loop, tracking total uploaded percent.

I'm guessing, based on a report, that this will go to 100, then on the next file attempt to reset to 0, which the API doesn't allow.

**Comment by 1NoOne1** [comment]
with recent code changes, it is now throwing an error message related to thread being restarted again. 
error output is posed at https://gist.github.com/1NoOne1/44164273e36485301d7c69da7a72cdcd .

**Comment by sivel** [comment]
> with recent code changes, it is now throwing an error message related to thread being restarted again.
> error output is posed at https://gist.github.com/1NoOne1/44164273e36485301d7c69da7a72cdcd .

I cannot reproduce this error when using an ovf/ova with multiple vmdks.  Your line numbers are off from the file in this PR, so I'm not sure how you added the changes, but I'd have to guess that there is a problem in your copy.  Specifically, there are likely 2 lines that read `uploader.start()`

**Comment by 1NoOne1** [comment]
Yes, sivel we are good on this. I copied lines wrongly. working like a charm.

**Comment by piotrplazienski** [comment]
Hi @sivel, that's great work.
For idempotency, maybe use approach from other vmware modules and introduce state option?
You could then use
_present_ - make sure it's there, do nothing if vm of such name exists
_redeployed_ - deploy, if it's there delete and deploy again
maybe use it instead of allow duplicates: there can be _state: present_allow_duplicates_ (maybe there's better name somewhere :) )
you can even throw in '_absent_' for completion.

Also, do you have any info about vcsim ova support? We would really like to use it, and I understand it blocks you from writing tests and merging. If there's some work left I am willing to help.

**Comment by dagwieers** [comment]
ovftool supports using HTTP URLs on the command-line, it would be nice if the module would support this too. https://www.vmware.com/support/developer/ovf/ovf420/ovftool-420-userguide.pdf

**Comment by sivel** [comment]
> ovftool supports using HTTP URLs on the command-line, it would be nice if the module would support this too.

HTTP URLs for the path to the OVF/OVA?

I *might* be able to do that.  Doing so would not be straight forward.  If it were an OVA, it would have to be downloaded fully to the "controller" first, since the API doesn't actually support OVA files directly, it requires OVF+VMDK and processed through `TarFileProgressReader`.

It may be easier in the short term to require the playbook to fetch the file first.  But I'll explore allowing HTTP URLs.

**Comment by dagwieers** [comment]
@sivel Sure. I also have an issue where the module is unable to find the datastore and  I cannot seem to specify a cluster or a folder either, which I will have to troubleshoot tomorrow. I really would like this to work so I can get rid of the complex role we build previously.

**Comment by sivel** [comment]
> I also have an issue where the module is unable to find the datastore and I cannot seem to specify a cluster or a folder either

I am writing this from a perspective of a person who has no access to anything other than a standalone esxi server.  So I can't comment on cluster capability.

I'm unsure about including an option for the folder, that is currently determined by `datacenter.vmFolder`.

Anything outside of my ability to personally test may have to wait until after this is merged.

As far as timing, this module may take a while to land, since we are waiting on vcsim support for OVF deploys, before merging this.

**Review by dericcrago — APPROVED** [review]
Just tested this, it works for me, just a couple of notes:
- VM folder placement would be a nice addition
- When setting `power_on: False`, needing to set `wait: False` or `wait_for_ip_address: False` seems unnecessary.

**Comment by sivel** [comment]
@dericcrago just as a heads up.  I've removed the unnecessary failure with power_on=False and wait_for=True

Additionally, I've added a folder arg, that defaults to `datacenter.vmFolder`.

**Comment by michaland** [comment]
@sivel 
> In my case the OVF is ubuntu-16.04-amd64.ovf and then it references ubuntu-16.04-amd64-disk1.vmdk

Is there some way to deploy OVF which points to more than one vmdk - more precisely: to one vmdk devided into e.g. 3 chunks?

e.g. 
R02.ovf
R02-disk1.vmdk.000000000
R02-disk1.vmdk.000000001
R02-disk1.vmdk.000000002

There is no problem with manual deploy via vSphere Client, but when I try to do it using vmware_deploy_ovf module I get:

FAILED! => {"changed": false, "msg": "Failed to find VMDK file at ...

**Comment by sivel** [comment]
@michaland I'd need to see the OVF file, and figure out how to build an ovf/vmdk that represented in that manner.  Offhand I'm unsure why it wouldn't work.

I hope to get this merged soon, and we can always work on extending functionality over time.

**Comment by Akasurde** [comment]
@sivel Thanks for the PR.

**Comment by judyQD** [comment]
thanks for your code, it did help me a lot. But I met a question when delpoy ovf using module vmware_deploy_ovf. I cannot add the DVS networks to the VM. Could you help have a look? Thank you in advance.

**Comment by albers** [comment]
Sadly this module didn't make it into 2.6.0. Will it be available in 2.6.1?

**Comment by Akasurde** [comment]
@albers I am afraid new modules are not backported to older version. :(

**Comment by albers** [comment]
I was hoping to get it in 2.6.0 because it was merged before 2.6.0 got released.
So I'll have to wait until 2.7.0?

**Comment by Akasurde** [comment]
It was merged after 2.6 feature freeze so couldn't make to 2.6. You can copy module directly to Ansible module installation directory (typically `<python_site_package>/lib/modules/cloud/vmware/`) or `library` directory of role.

**Comment by albers** [comment]
@Akasurde Thanks for clarifying. I'll try copying the module.

**Comment by albers** [comment]
@Akasurde I was able to successfully use the new module after copying the module file over Ansible 2.6.1 as you suggested. But I also had to replace  `module_utils/vmware.py` with the current version. The module worked perfectly.

**Comment by Akasurde** [comment]
@albers thanks for the feedback.

**Comment by albers** [comment]
@sivel, @Akasurde  FYI, I created a role that uses `vmware_deploy_ovf`: https://galaxy.ansible.com/albers/vmware_ubuntu_cloud_image

**Comment by SumitJaiswal997** [comment]
Trying to build a VM with OVF stored in Ansible control server, required ports are opened. When I use the vmware_deploy_ovf module if doesn't locate the datastore. It works when vmware_guest module is used, this uses template to build VM.  I am using Datastore cluster name but I have tried individual datastore name also. DRS is enabled in the VMWare setup. Please suggest. Thanks
Error 
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Datacentername could not be located"} 
-----------------------
- hosts: localhost
  connection: local
  tasks:

  - vmware_deploy_ovf:
     hostname: vcenter.test.com
     username: test\uerid
     password: Password@q9223
     validate_certs: no
     ova: /var/www/html/windows_ovf/Test2.ovf
     datacenter: datacentername
     cluster: clustename
     folder: /datacentername/vm
     datastore: datastorename
     name: AutomationVMovf
     wait_for_ip_address: false
  #delegate_to: localhost
----------------------------

**Pull request merged**
